### ISSUE TITLE
Make Action Serialization Opt-In

### ISSUE THREAD

**Issue by Ben-G** [Swift-Flow/Swift-Flow#12]
Currently the library requires all actions to be serializable. After brief discussion with @russbishop I would like to follow his suggestion and make them opt-in.

This will have the downside that some developers will not immediately understand that developer tools (time traveling) won't work unless all actions are serializable, but we can work with warnings, log messages, etc. to communicate that.

Many existing Obj-C/Swift frameworks provide objects that are hard to serialize and therefor could not easily be used within actions.

**Comment by Ben-G** [comment]
This would also lower the priority for #2.

**Comment by Ben-G** [comment]
Some current thoughts on this issue:

- With current approach all parts of Swift Flow deal with `ActionType` which is a simple, standard action. The developer is responsible for creating typed actions from untyped actions wherever necessary. This makes deserialization easier; since we always deal with plain actions it doesn't matter if it's a deserialized one, or one that resided in memory
- If we'd naively allow `Any` type to be an action, we could have situations where the application relies on typed actions; that works as long the application has all actions in memory. Once you would serialize and deserialize these actions, they would come back as standard `ActionType` actions. Without changing any other code, you would end up with 2 code paths in reducers, one where you deal with the typed in-memory action and one where you deal with the plain deserialized one; that's obviously undesirable
- To avoid this, deserialization should include instantiating the typed action. That way reducers would always received typed actions the same way, independent of whether they were stored in memory or on disk. We would need to flag the serialized standard action to indicate that it is a typed action that requires instantiating a specific type upon deserialization. We would also need a mechanism to map a type string to a Swift type

**Comment by Ben-G** [comment]
This whole thing also relates to #8

**Comment by Ben-G** [comment]
We could do the dynamic instantiation of types using a mapping from a type string to a swift type:

```swift
let typemap: [String : StandardActionConvertible.Type] = ["MyCustomAction": MyCustomAction.self]
```

Full demonstration in this [API playground](https://github.com/Swift-Flow/API-Playgrounds/blob/master/StrongerTypedActions/ben-g/TypedActionSerialization.playground/Contents.swift).

**Comment by Ben-G** [comment]
This has been implemented as part of #16
